xds/csds/csds.go

```diff
@@ -0,0 +1,321 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package csds implements features to dump the status (xDS responses) the
+// xds_client is using.
+//
+// Notice: This package is EXPERIMENTAL and may be changed or removed in a later
+// release.
+package csds
+
+import (
+	"context"
+	"io"
+	"runtime"
+	"time"
+
+	v3adminpb "github.com/envoyproxy/go-control-plane/envoy/admin/v3"
+	v2corepb "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3statuspb "github.com/envoyproxy/go-control-plane/envoy/service/status/v3"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/xds/internal/client"
+	"google.golang.org/grpc/xds/internal/client/bootstrap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	_ "google.golang.org/grpc/xds/internal/client/v2" // Register v2 xds_client.
+	_ "google.golang.org/grpc/xds/internal/client/v3" // Register v3 xds_client.
+)
+
+// xdsClientInterface contains methods from xdsClient.Client which are used by
+// the server. This is useful for overriding in unit tests.
+type xdsClientInterface interface {
+	DumpLDS() (string, map[string]client.UpdateWithMD)
+	DumpRDS() (string, map[string]client.UpdateWithMD)
+	DumpCDS() (string, map[string]client.UpdateWithMD)
+	DumpEDS() (string, map[string]client.UpdateWithMD)
+	BootstrapConfig() *bootstrap.Config
+	Close()
+}
+
+var (
+	logger       = grpclog.Component("xds")
+	newXDSClient = func() (xdsClientInterface, error) {
+		return client.New()
+	}
+)
+
+type clientStatusServer struct {
+	// xdsClient will always be the same in practise. But we keep a copy in each
+	// server instance for testing.
+	xdsClient    xdsClientInterface
+	xdsClientErr error
+}
+
+// NewClientStatusDiscoveryServer returns an implementation of the CSDS server that can be
+// registered on a gRPC server.
+func NewClientStatusDiscoveryServer() v3statuspb.ClientStatusDiscoveryServiceServer {
+	ret := &clientStatusServer{}
+	xdsC, err := newXDSClient()
+	if err != nil {
+		ret.xdsClientErr = err
+		// Log error instead of returning an error, so that
+		// `NewClientStatusDiscoveryServer` is easier to use.
+		//
+		// Besides, if client cannot be created, the users should have received
+		// the error, because their client/server cannot work, either.
+		logger.Errorf("failed to create xds client: %v", ret.xdsClientErr)
```

Comment:
Also, `return` from inside the `if`, so that the code under `else` need not be indented.

Reply:
Done